### ISSUE TITLE
Support for 'MaxAllowed' number of services on a given node

### ISSUE THREAD

**Issue by cdancy** [coreos/fleet#1262]
Within the service files I'd like to be able to do something like this:

```
[X-Fleet]
MachineMetadata=location=chicago
Conflicts=monitor*
MaxAllowed=5
```

The 'MaxAllowed' is what I've added and wondering if we could add something like this to fleet. The idea is that no more than 'MaxAllowed' units can execute on the given selected node. Thoughts?

**Comment by wuqixuan** [comment]
I think it's useful. Have you already implemented it ? If not, we can help to do it.

**Comment by cdancy** [comment]
I've not coded any implementation just putting the idea out to see what others think. From looking at the code it doesn't *seem* too bad to implement but I'm admittedly not as familiar with fleet source as others here. If someone can take this one that would be great. The name is something I through to get the point across. If something else fits better so be it.

**Comment by wuqixuan** [comment]
I think the idea is same as the replication conception in Kubernetes. So it's meaningful in fleet also. We will help to implement and contribute the patch.

**Comment by Dynom** [comment]
Is the snippet you presented part of the config file of a unit? If so, isn't it a bit awkward to place a machine limit in a unit file?

**Comment by wuqixuan** [comment]
The snippet is in part of the unit file. It identify how much the unit can be lauched in how many machines. We can think of it as a machine limit. But I think it's meaningful.

**Comment by philk** [comment]
We've regularly seen fleet schedule all units to a single host (when doing something like `fleetctl start my@{1,2,3,4,5}.service`) and this looks like a really nice way to work around that without being extremely restrictive and using a direct `Conflicts=` line

**Comment by Dynom** [comment]
Ahh, now I understand. So specifying `MaxAllowed=2` in the example of @philk would result in at most 2 services on any given node. Didn't look at it from that angle. That sounds very useful indeed.

**Comment by bcwaldon** [comment]
I don't think adding MaxAllowed to a given unit is the right thing to do here. Either you want to allow a maximum number of units onto each machine, or you want to spread out a given type of unit based on some criteria.

**Comment by Dynom** [comment]
A perhaps better solution would be for Fleet to understand the load-impact of a unit. But I recon this is exceptionally hard to determine. To protect machines the `MaxAllowed` directive could be a nice work around to mimic that behaviour, albeit it being rather primitive.

In one particular situation I now do the following:
* Schedule a machine with HAProxy
* Schedule N web-units to the machine with the HAProxy unit
* HAProxy figures out how many web-units there are and load balances requests to the N web-units

But if I spawn 7 web-units over two HAProxy machines, there is an uneven balance. If you have 80 units running in a cluster, it's easy to miss one. This directive will result in an (infamous) administrative error message from Fleet indicating that at least something is afoot.

I think there is merit to this feature request, but perhaps not in the suggested form.

**Comment by cdancy** [comment]
Load balancing, and all that it entails, is quite a different thing than asking fleet to make sure no more than X number of the same unit are scheduled on the same machine. This request, however primitive, seems to fall inline with other X-Fleet parameters already available.

**Comment by bcwaldon** [comment]
@cdancy how are you defining that two given units are the "same"?

**Comment by cdancy** [comment]
Perhaps this is where the confusion is coming into play. My assumption is that this would be based on everything to the left-hand side of the '@' sign. For example if user starts the hello-world@.service he/she could do so at most 'MaxAllowed' (or whatever name we want to use) times on a given node before being rejected. This would also include templated units.

Looking at the source it seems this feature would be applied last after all other filters have completed (which is why I listed it last above if only to make the point). So if after all filters applied fleet deems hello-world@.service is to run on node X it would then query said node to see how many 'hello-world' units are running whether templated or not. If scheduling the unit pushes us past MaxAllowed then we get an error back saying as much.

**Comment by berglh** [comment]
MaxAllowed would be extremely useful.

In my instance, I have a unit file that changes its parameters based on the number of instances already running on a particular host by querying etcd for the current state. The unit files are for launching elasticsearch nodes that have persistent data storage locally attached.

Currently, the physical infrastructure can only support a maximum of three instances of this unit file. In a cluster of 5 machines, this is a problem when fleet tries to balance depending on the available hosts, then I have unsupported unit files running when the number of instances is >3 on a fleet member.

I have written the unit file with future scalability in mind, even though I do not have the resources for it currently. With the MaxAllowed setting, I can just increase the number as resources come to hand and ensure I don't end up running too many instances whereby the additional instances attempt to attach to storage that isn't there.

The current solution is to have three separately name unit files and use the Conflict pattern matching to restrict them. When I want to scale up, I would then have to make more unit files. This would avoid that requirement.

**Comment by berglh** [comment]
Can someone point us in the right direction to help writing this feature? It would really solve a lot of problems for our orchestration where we want the flexibility of being able to scale out if needed (by destroying and recreating), but really need to be able to set MaxAllowed to prevent overallocation of a unit file type on any one host (due to do hardware limitations). I've been starting to dabble in go, maybe I can make a contribution here. Where are the current conflicts/filter functions?

**Comment by berglh** [comment]
It's been a while since any discussion on this topic, so I thought I'd poke the coals. Is there any other way in recent versions to accomplish a MaxAllowed number of units per fleet node? It's really painful having to manage this manually and to prevent the re-allocation of units to other hosts while one is just rebooting.

I'd be happy to get my hands dirty, but not sure where to started to implement such a feature, I'm assuming we could implement this as an extension of the Conflicts functions.

**Comment by Dynom** [comment]
I'm still thinking about (and looking for) a more advanced scheduling construction. I think specifying a `MaxAllowed` directive requires you to know too much about the unit, in combination with the machines they will land on. Plus it feels like a work around.

I'm not sure how feasible it is to think in terms of hardware specifications, but what if it was possible to describe the unit you're trying to load and let the scheduler make the decisions?

A very simplistic unit definition could be:
```ini
[X-Fleet]
Conflicts=%p@*

PlacementMemMax=512M
PlacementIOPSMax=900
PlacementCPUCores=1
...

[Service]
...
```

I deliberately prefix the directives with `Placement` to indicate to the author that it's about placement and not actual resource monitoring (re-)scheduling. Omitted Placement directives means they shouldn't be used to decide placement, this way one can e.g. only specify memory requirements (no one likes OOM kills).

However, specifying these indicators allows the scheduler, knowing what specifications each machine in a cluster has, to allocate based on resources. What this solves is very basic matching of resources. It doesn't take anything complex into account. But it does allow for more granular placing of units, including descriptive errors when invalid units are to be scheduled.

I work around this problem by `X-Conflict`-ing similarly resource requiring unit's, but that is too limiting and requires a lot of administration.

-- 2 cents

**Comment by berglh** [comment]
@Dynom 

>  I think specifying a MaxAllowed directive requires you to know too much about the unit, in combination with the machines they will land on. Plus it feels like a work around.

Perhaps a work around for your specific problem. I do appreciate the ability to schedule units based on resources if you have a large and varied hardware pool; but it is solving a different problem to what I'm faced with.

My issue is that one of my unit file has a specific requirement to run x number time on each fleet member server. When initiating those instances they are distributed evenly across the fleet. That unit in it's operation is not dynamic enough to be rescheduled to another server and operate when an x number is already scheduled on each server, it will actually just fail to run if it runs more than x times due to the underlying storage. Once it's been rescheduled it then does not move back to the original server when it becomes available, meaning I've lost that unit and manual intervention is required to return those units to the original server.

As much as I strive for scalability and agnosticism with the unit files in use, I can't always change the underlaying software to work in the way I would like. 

I'm quite adamant that having a `MaxAllowed` feature is really the only the solution to my problem other then creating x differently named units files that contain the exact same contents and avoid the re-allocation of those units by using the `X-Conflicts` setting.

I think we are faced with two separate issues that require different solutions to solve them :+1:

**Comment by antrik** [comment]
@berglh I fail to see how your use case is fundamentally different. As I understand it you also have a limited resource (the storage in your case) that dictates scheduling requirements.

Having said that, I don't know the right solution for this either. Like many other issues, it goes back to the fact that fleet presently doesn't have -- nor pretends to have -- much in the way of an actual scheduler...

**Comment by berglh** [comment]
Just an update to this request. 

It turns out I was able to get around this particular requirement by creating three symlinks to the same unit file. By using different link names and restricting to one instance per machine, I was able to simulate a MaxAllowed function. Using this method, I still only have one unit file to maintain for the whole cluster, and the name of the symlink dictates the role for that unit file.

I also matched these with a service-discovery unit instance number using the MachineOf directive to prevent it being migrated across to other hosts upon host restart or failure. When the host came back online, it was assigned the same units and were subsequently matched to their storage. As I have other teams maintaining patching on these boxes, this has been crucial for minimal downtime for physical nodes.

While this does not give me the usefulness of MaxAllowed directive where I would no longer require symlinks to a common unit file, I was able to get past my particular issue by considering all the functions of the x fleet directives. Yes, it's less than ideal, but a mechanism for me to solve my particular problem  was there.

I would still love to see this feature, particularly as some solutions will be tied to host specific storage but the ability to scheduled common units across a cluster of hosts using fleet and CD/CI frameworks is useful.
